### PR TITLE
Implement XRP DEX order parsing -> EdgeTxAction's

### PR DESCRIPTION
**Pull request by Jon-edge**
### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

https://github.com/EdgeApp/edge-core-js/pull/567

### Description

Parse XRP DEX orders into EdgeTxActions. Supports OfferCreate and OfferCancel type order tx's

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205412691849558

### CODE REVIEW

File: src/ripple/RippleEngine.ts [hunk 1]
```diff
@@ -275,6 +303,155 @@ export class XrpEngine extends CurrencyEngine<
     }
   }
 
+  /**
+   * Parse TakerGets or TakerPays into an EdgeAssetAmount
+   * */
+  parseRippleDexTxAmount = (
```
Comment (paullinator) [inline src/ripple/RippleEngine.ts:309]: Replace most of parseRippleDexTxAmount and processRippleDexTx with getBalances to determine src/dest amounts of order fulfillments.

**Review by paullinator — CHANGES_REQUESTED** [review]

File: src/ripple/RippleEngine.ts (hunk 1 above)
Comment (Jon-edge) [reply src/ripple/RippleEngine.ts:309]: The way I had it done is more accurate and complete according to [xrpscan.com](http://xrpscan.com/). 

The balances object is good for including mainnet fees in the xrp amount total, but for some reason underreports the actual amounts involved in the trades (regardless of fees - token amounts are underreported). Also there's no way to tell unfilled order amounts from the `Balances` obj either.

Comment (Jon-edge) [reply src/ripple/RippleEngine.ts:309]: Comparing against xrpscan.com shows alignment between my amounts and the summary amounts given by api queries to xrp explorers

File: src/ripple/RippleEngine.ts [hunk 2]
```diff
@@ -275,6 +303,155 @@ export class XrpEngine extends CurrencyEngine<
...
+      // Order was fully filled
+      const isFullyFilled = hasDeletedNodes && !isOpenOrder
+
+      // Don't care about partial fills - counting them as general fills
+      type =
+        isFullyFilled || isPartiallyFilled ? 'swapOrderFill' : 'swapOrderPost'
+
+      // Parse amounts
+      const { TakerPays, TakerGets } = tx
+      sourceAsset = this.parseRippleDexTxAmount(TakerGets)
```
Comment (paullinator) [inline src/ripple/RippleEngine.ts:409]: This always get the amounts from the `tx` object. But in the case of a partial order fulfillment, the correct amounts are in the Nodes.

**Review by paullinator — CHANGES_REQUESTED** [review]

**Review by paullinator — APPROVED** [review]

File: src/ripple/RippleEngine.ts (hunk 2 above)
Comment (Jon-edge) [reply src/ripple/RippleEngine.ts:409]: We don't actually display the amounts from here at this time. 

This implementation is tracking 2 things consistently across all order states: the state of the order, and the posted order amount. 

Fill amount is a completely separate thing and I think is a mistake to conflate into the `source/destAsset` amounts described here.

Also if we only track fill amount, we actually are mis-reporting later down the road when the order status changes. We won't have knowledge of that unless we go back and modify related transactions as we process newer transactions. Imo, fill amount should be a completely separate field.

At this point in time, the only thing we are certain of is the total amounts of the original offer. Even if we have a partial fill immediately upon posting, we don't know if a later tx from someone else further modified the fill amount.

File: src/ripple/RippleEngine.ts [hunk 3]
```diff
@@ -275,6 +303,155 @@ export class XrpEngine extends CurrencyEngine<
...
+    const {
+      currency,
+      issuer,
+      value
+      // Taker pays/gets XRP if 'TakerPays/Gets' is a plain string
+    } =
+      typeof takerAmount === 'string'
+        ? { currency: 'XRP', issuer: undefined, value: takerAmount }
+        : takerAmount
+    const isTakerToken = currency !== 'XRP' && issuer != null
```
[the remark this answers is not in the document]
Comment (Jon-edge) [reply src/ripple/RippleEngine.ts:321]: I'll add the assert, but we still need to keep the entire `isTakerToken` statement, or else `makeTokenId` right after complains

**Pull request merged**